### PR TITLE
PWGHF:B+ Creator and B+ task - no selector yet

### PR DESCRIPTION
**Pull request by apalasciano**
First Implementation of B+->D0(bar)+ pi+ Creator, B task, CMakeFiles.txt and HFSecondaryVertex.h

### CODE REVIEW

**Comment by vkucera** [comment]
There is some problem with this PR as it seems to revert previous merged PRs.
Please fix it.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file HFCandidateCreator2Prong.cxx
+/// \brief Reconstruction of heavy-flavour 2-prong decay candidates
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx:12]: Update

File: Analysis/Tasks/PWGHF/taskBPlus.cxx [hunk 2]
```diff
@@ -33,48 +35,131 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
 
 #include "Framework/runDataProcessing.h"
 
-/// B+ analysis task
-struct TaskBplus {
+/// BPlus analysis task
+struct TaskBPlus {
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskBPlus.cxx:39]: Why?

File: Analysis/Tasks/PWGHF/taskBPlus.cxx [hunk 3]
```diff
@@ -33,48 +35,131 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
...
+      registry.fill(HIST("hImpParErr"), candidate.errorImpactParameter0());
+      registry.fill(HIST("hImpParErr"), candidate.errorImpactParameter1());
+      registry.fill(HIST("hDecLenErr"), candidate.errorDecayLength());
+      registry.fill(HIST("hDecLenXYErr"), candidate.errorDecayLengthXY());
+    } // candidate loop
+  }   // process
+};    // struct
 
-  Filter filterSelectCandidates = (aod::hf_selcandidate_d0::isSelD0 >= d_selectionFlagD0 || aod::hf_selcandidate_d0::isSelD0bar >= d_selectionFlagD0bar);
+struct TaskBPlusMC {
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskBPlus.cxx:86]: MIssing documentation

File: Analysis/Tasks/PWGHF/taskBPlus.cxx [hunk 4]
```diff
@@ -33,48 +35,131 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
...
+  HistogramRegistry registry{
+    "registry",
+    {{"hPtRecSig", "2-prong candidates (rec. matched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtRecBg", "2-prong candidates (rec. unmatched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGen", "2-prong candidates (gen. matched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGenProng0", "2-prong candidates (gen. matched);prong 0 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGenProng1", "2-prong candidates (gen. matched);prong 1 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGenSig", "2-prong candidates (rec. matched);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hCPARecSig", "2-prong candidates (rec. matched);cosine of pointing angle;entries", {HistType::kTH1F, {{110, -1.1, 1.1}}}},
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskBPlus.cxx:95]: Fix the titles and labels where needed (`2-prong candidates` -> `MC particles` for gen. values, distinction rec. vs. gen. for axis labels). See the D0 task.

File: Analysis/Tasks/PWGHF/taskBPlus.cxx [hunk 5]
```diff
@@ -33,48 +35,131 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
...
+     {"hEtaRecBg", "2-prong candidates (rec. unmatched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hEtaGen", "2-prong candidates (gen. matched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hMassRecSig", "2-prong candidates (rec. matched);inv. mass (D0bar #pi+) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 10.}}}},
+     {"hMassRecBg", "2-prong candidates (rec. unmatched);inv. mass (D0bar #pi+) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 10.}}}},
+     {"hDecLengthRecSig", "2-prong candidates (rec. matched);decay length (cm);entries", {HistType::kTH1F, {{200, 0., 2.}}}},
+     {"hDecLengthRecBg", "2-prong candidates (rec. unmatched);decay length (cm);entries", {HistType::kTH1F, {{200, 0., 2.}}}}}};
 
-  Partition<aod::BigTracks> positiveTracks = aod::track::signed1Pt >= 0.f;
+  // Configurable<int> selectionFlagB{"selectionFlagB", 1, "Selection Flag for B"};
+  Configurable<double> cutYCandMax{"cutYCandMax", -1., "Upper cut rap."};
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskBPlus.cxx:106]: `"max. cand. rapidity"`

File: Analysis/Tasks/PWGHF/taskBPlus.cxx [hunk 6]
```diff
@@ -33,48 +35,131 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
...
+        registry.fill(HIST("hPtRecSig"), candidate.pt());
+        registry.fill(HIST("hCPARecSig"), candidate.cpa());
+        registry.fill(HIST("hEtaRecSig"), candidate.eta());
+        registry.fill(HIST("hDecLengthRecSig"), candidate.decayLength());
+        registry.fill(HIST("hMassRecSig"), InvMassBplus(candidate));
+        registry.fill(HIST("hPtRecBg"), candidate.pt());
+        registry.fill(HIST("hCPARecBg"), candidate.cpa());
+        registry.fill(HIST("hEtaRecBg"), candidate.eta());
+        registry.fill(HIST("hDecLengthRecBg"), candidate.decayLength());
+        registry.fill(HIST("hMassRecBg"), InvMassBplus(candidate));
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskBPlus.cxx:133]: You're filling background histograms with signal. Why?

File: Analysis/Tasks/PWGHF/taskBPlus.cxx [hunk 7]
```diff
@@ -33,48 +35,131 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
...
+    // MC gen. level
+    //Printf("MC Particles: %d", particlesMC.size());
+    for (auto& particle : particlesMC) {
+      if (std::abs(particle.flagMCMatchGen()) == 1 << BPlusToD0Pi) {
+        registry.fill(HIST("hPtGen"), particle.pt());
+        float ptProngs[2];
+        int counter = 0;
+        for (int iD = particle.daughter0(); iD <= particle.daughter1(); ++iD) {
+          ptProngs[counter] = particlesMC.iteratorAt(iD).pt();
+          counter++;
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskBPlus.cxx:146]: I don't understand this. What is it supposed to do?

File: Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+      //Printf("New rec. candidate");
+      flag = 0;
+      auto D0barTrack = candidate.index0();
+      auto arrayDaughters = array{candidate.index1_as<aod::BigTracksMC>(), D0barTrack.index0_as<aod::BigTracksMC>(), D0barTrack.index1_as<aod::BigTracksMC>()};
+
+      // B± → D0bar(D0) π± → (K± π∓) π±
+      //Printf("Checking B± → D0(bar) π±");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, 521, array{321, -kPiPlus, +kPiPlus}, true, &sign, 2);
+      if (indexRec > -1) {
+        flag = sign * (1 << BPlusToD0Pi);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx:240]: Use the full identifier of `BPlusToD0Pi`.

File: Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    // Match reconstructed candidates.
+    for (auto& candidate : candidates) {
+      //Printf("New rec. candidate");
+      flag = 0;
+      auto D0barTrack = candidate.index0();
+      auto arrayDaughters = array{candidate.index1_as<aod::BigTracksMC>(), D0barTrack.index0_as<aod::BigTracksMC>(), D0barTrack.index1_as<aod::BigTracksMC>()};
+
+      // B± → D0bar(D0) π± → (K± π∓) π±
+      //Printf("Checking B± → D0(bar) π±");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, 521, array{321, -kPiPlus, +kPiPlus}, true, &sign, 2);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx:238]: Use named PDG values.

Comment (vkucera) [reply Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx:238]: Where do you check D0(bar) → π± K∓?

File: Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+      // B± → D0bar(D0) π± → (K± π∓) π±
+      //Printf("Checking B± → D0(bar) π±");
+      if (RecoDecay::isMatchedMCGen(particlesMC, particle, 521, array{-421, +kPiPlus}, true, &sign, 1)) {
+        // D0(bar) → π± K∓
+        //Printf("Checking D0(bar) → π± K∓");
+        std::vector<int> arrayDaughterB;
+        RecoDecay::getDaughters(particlesMC, particle, &arrayDaughterB, array{421}, 1); //it takes abs. of PDG codes
+        auto D0candidateMC = particlesMC.iteratorAt(arrayDaughterB[0]);
+        if (RecoDecay::isMatchedMCGen(particlesMC, D0candidateMC, 421, array{-321, +kPiPlus}, true, &sign)) {
+          flag = (-1) * sign * (1 << BPlusToD0Pi);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx:259]: Why did you flip the signs?

**Comment by vkucera** [comment]
@apalasciano Any updates?

**Comment by vkucera** [comment]
@apalasciano So this PR will be replaced by another one, right?

**Comment by apalasciano** [comment]
Yes, it is. 
The new is this one  #6265

**Pull request closed without merging**